projects/kit/components/calendar-range/calendar-range.component.ts

```diff
@@ -208,7 +208,41 @@ export class TuiCalendarRange implements OnChanges {
             const inDisabledRange =
                 disabledBefore.dayBefore(item) && disabledAfter.dayAfter(item);
 
-            return inDisabledRange || disabledItemHandler(item);
+            return (
+                inDisabledRange || this.isDisabledItem(disabledItemHandler, value, item)
+            );
         };
     }
+
+    private isDisabledItem(
+        disabledItemHandler: TuiBooleanHandler<TuiDay>,
+        value: TuiDayRange | null,
+        item: TuiDay,
+    ): boolean {
+        return (
+            disabledItemHandler(item) ||
+            (!!value?.isSingleDay &&
+                this.isExternalDaysDisabled(disabledItemHandler, value.from, item))
+        );
+    }
+
+    private isExternalDaysDisabled(
+        disabledItemHandler: TuiBooleanHandler<TuiDay>,
+        startDay: TuiDay,
+        item: TuiDay,
+    ): boolean {
+        let temp = item;
+
+        while (temp.dayBefore(startDay) || temp.dayAfter(startDay)) {
+            if (disabledItemHandler(temp)) {
```

Comment:
Disabled item handler is called on each change detection on each available day, this function runs in creating classes in a loop for each day in between, I fear it might be detrimental to performance. Can we optimize it somehow? Can we measure how much of an impact this change brings?

Reply:
changed search logic. After first day selection we search range of available dates. Than for each item we check its availability if item contains in selected range